Empty slo-exporter configuration is declared as valid

## Problem

Empty main configuration file is according to config checker valid. This is incorrect.

## Reproducibility 100%
```sh
$ ./slo_exporter --version
slo_exporter version 6.4.0 (from commit 3c44f001e863265e588e319138e743bc682b70b3 at 2020-08-07T10:58:03Z by goreleaser)

$ touch /tmp/slo-exporter.yaml
✓ freznicek@localhost.localdomain /tmp

$ ./slo_exporter --config-file /tmp/slo-exporter.yaml --check-config
time="2020-08-07T16:23:38.01516+02:00" level=info msg="Configuration is valid!"
✓ freznicek@localhost.localdomain /tmp

$ ./slo_exporter --config-file /tmp/slo-exporter.yaml
time="2020-08-07T16:23:42.81021+02:00" level=warning msg="changing readiness status to not ok, reason: initializing" component=prober
...
time="2020-08-07T16:23:42.8119+02:00" level=info msg="changing readiness status to ok" component=prober
time="2020-08-07T16:23:43.81233+02:00" level=info msg="finished processing all logs"
time="2020-08-07T16:23:43.81245+02:00" level=info msg="gracefully shutting down"
time="2020-08-07T16:23:43.8125+02:00" level=warning msg="changing readiness status to not ok, reason: shutting down" component=prober
time="2020-08-07T16:23:43.81258+02:00" level=info msg="see you next time!"
panic: runtime error: index out of range [0] with length 0

goroutine 1 [running]:
github.com/seznam/slo-exporter/pkg/pipeline.(*Manager).StopPipeline(0xc00025e240, 0xe85e60, 0xc000316180, 0xe85e60)
        /home/circleci/project/pkg/pipeline/manager.go:69 +0xc1
main.main()
        /home/circleci/project/cmd/slo_exporter.go:224 +0x1365
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+
+## [v6.4.1] 2020-08-25
 ## Fixed
 - [#27](https://github.com/seznam/slo-exporter/pull/27) Empty configuration is now evaluated as invalid
 
@@ -24,7 +26,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - [#18](https://github.com/seznam/slo-exporter/pull/18) Tailer: Upgraded hpcloud/tail package to latest commit.
 
 ## Added
-- [#20](https://github.com/seznam/slo-exporter/pull/20) CI: 
+- [#20](https://github.com/seznam/slo-exporter/pull/20) CI:
      - Release additional Docker image tags:
         - `latest` Latest released version.
         - `vX` Latest released version for the major version.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,7 +6,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 ## Fixed
-- [#26](https://github.com/seznam/slo-exporter/pull/26) Empty configuration is now evaluated as invalid
+- [#27](https://github.com/seznam/slo-exporter/pull/27) Empty configuration is now evaluated as invalid
 
 ## [v6.4.0] 2020-08-07
 ## Added

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+## Fixed
+- [#26](https://github.com/seznam/slo-exporter/pull/26) Empty configuration is now evaluated as invalid
 
 ## [v6.4.0] 2020-08-07
 ## Added

--- a/cmd/slo_exporter.go
+++ b/cmd/slo_exporter.go
@@ -199,8 +199,10 @@ func main() {
 	}
 	pipelineManager.RegisterWebInterface(router)
 
-	// Start the pipeline items `processing
-	pipelineManager.StartPipeline()
+	// Start the pipeline processing
+	if err := pipelineManager.StartPipeline(); err != nil {
+		logger.Fatalf("failed to start the pipeline: %v", err)
+	}
 
 	// listen for OS signals
 	sigChan := make(chan os.Signal, 3)

--- a/pkg/pipeline/manager.go
+++ b/pkg/pipeline/manager.go
@@ -28,6 +28,9 @@ func NewManager(moduleFactory moduleFactoryFunction, config *config.Config, logg
 		pipeline: []pipelineItem{},
 		logger:   logger,
 	}
+	if len(config.Pipeline) < 1 {
+		return nil, fmt.Errorf("failed to create the pipeline as no modules defined in the config")
+	}
 	// Initialize the pipeline and link it together.
 	for _, moduleName := range config.Pipeline {
 		newPipelineItem, err := manager.newPipelineItem(moduleName, config, moduleFactory)

--- a/pkg/pipeline/manager.go
+++ b/pkg/pipeline/manager.go
@@ -6,8 +6,8 @@ import (
 	"github.com/gorilla/mux"
 	"github.com/iancoleman/strcase"
 	"github.com/prometheus/client_golang/prometheus"
-	"github.com/sirupsen/logrus"
 	"github.com/seznam/slo-exporter/pkg/config"
+	"github.com/sirupsen/logrus"
 	"strings"
 	"time"
 )
@@ -59,14 +59,14 @@ func (m *Manager) StartPipeline() error {
 	if len(m.pipeline) == 0 {
 		return fmt.Errorf("failed to execute the empty pipeline (no pipeline modules defined in the config)")
 	}
-    var pipelineSchema []string
+	var pipelineSchema []string
 	for _, pipelineItem := range m.pipeline {
 		pipelineItem.module.Run()
 		pipelineSchema = append(pipelineSchema, pipelineItem.name)
 	}
 	m.logger.Info("pipeline schema: " + strings.Join(pipelineSchema, " -> "))
 	m.logger.Info("pipeline started")
-  return nil
+	return nil
 }
 
 func (m *Manager) StopPipeline(ctx context.Context) chan struct{} {

--- a/pkg/pipeline/manager.go
+++ b/pkg/pipeline/manager.go
@@ -28,9 +28,6 @@ func NewManager(moduleFactory moduleFactoryFunction, config *config.Config, logg
 		pipeline: []pipelineItem{},
 		logger:   logger,
 	}
-	if len(config.Pipeline) < 1 {
-		return nil, fmt.Errorf("failed to create the pipeline as no modules defined in the config")
-	}
 	// Initialize the pipeline and link it together.
 	for _, moduleName := range config.Pipeline {
 		newPipelineItem, err := manager.newPipelineItem(moduleName, config, moduleFactory)
@@ -56,16 +53,20 @@ type Manager struct {
 	logger   logrus.FieldLogger
 }
 
-func (m *Manager) StartPipeline() {
+func (m *Manager) StartPipeline() error {
 	m.logger.Info("starting pipeline... ")
-	var pipelineSchema []string
+
+	if len(m.pipeline) == 0 {
+		return fmt.Errorf("failed to execute the empty pipeline (no pipeline modules defined in the config)")
+	}
+    var pipelineSchema []string
 	for _, pipelineItem := range m.pipeline {
 		pipelineItem.module.Run()
 		pipelineSchema = append(pipelineSchema, pipelineItem.name)
 	}
 	m.logger.Info("pipeline schema: " + strings.Join(pipelineSchema, " -> "))
 	m.logger.Info("pipeline started")
-
+  return nil
 }
 
 func (m *Manager) StopPipeline(ctx context.Context) chan struct{} {

--- a/pkg/pipeline/manager_test.go
+++ b/pkg/pipeline/manager_test.go
@@ -41,8 +41,16 @@ func (t *testModule) OutputChannel() chan *event.Raw {
 	return make(chan *event.Raw)
 }
 
-func newTestManager() (*Manager, error) {
+func newEmptyManager() (*Manager, error) {
 	manager, err := NewManager(testModuleFactory, &config.Config{Pipeline: []string{},}, logrus.New())
+	if err != nil {
+		return nil, err
+	}
+	return manager, nil
+}
+
+func newTestManager() (*Manager, error) {
+	manager, err := newEmptyManager()
 	if err != nil {
 		return nil, err
 	}
@@ -68,6 +76,15 @@ test_module_test 0
 
 	err = testutil.GatherAndCompare(registry, strings.NewReader(expectedMetrics))
 	assert.NoError(t, err)
+}
+
+
+// Empty pipeline refuses to start
+func TestManager_StartEmptyPipeline(t *testing.T) {
+	manager, err := newEmptyManager()
+	assert.NoError(t, err)
+	err = manager.StartPipeline()
+	assert.Error(t, err)
 }
 
 func TestManager_StartPipeline(t *testing.T) {

--- a/pkg/pipeline/manager_test.go
+++ b/pkg/pipeline/manager_test.go
@@ -4,11 +4,11 @@ import (
 	"context"
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/client_golang/prometheus/testutil"
+	"github.com/seznam/slo-exporter/pkg/config"
+	"github.com/seznam/slo-exporter/pkg/event"
 	"github.com/sirupsen/logrus"
 	"github.com/spf13/viper"
 	"github.com/stretchr/testify/assert"
-	"github.com/seznam/slo-exporter/pkg/config"
-	"github.com/seznam/slo-exporter/pkg/event"
 	"strings"
 	"testing"
 )
@@ -42,7 +42,7 @@ func (t *testModule) OutputChannel() chan *event.Raw {
 }
 
 func newEmptyManager() (*Manager, error) {
-	manager, err := NewManager(testModuleFactory, &config.Config{Pipeline: []string{},}, logrus.New())
+	manager, err := NewManager(testModuleFactory, &config.Config{Pipeline: []string{}}, logrus.New())
 	if err != nil {
 		return nil, err
 	}
@@ -77,7 +77,6 @@ test_module_test 0
 	err = testutil.GatherAndCompare(registry, strings.NewReader(expectedMetrics))
 	assert.NoError(t, err)
 }
-
 
 // Empty pipeline refuses to start
 func TestManager_StartEmptyPipeline(t *testing.T) {

